test_conformance/subgroups/test_subgroup_ballot.cpp

```diff
@@ -0,0 +1,1089 @@
+    "__kernel void test_get_sub_group_gt_mask(const __global Type *in, "
+    "__global int4 *xy, __global Type *out)\n"
+    "{\n"
+    "    int gid = get_global_id(0);\n"
+    "    XY(xy,gid);\n"
+    "    xy[gid].z = get_max_sub_group_size();\n"
+    "    Type x = in[gid];\n"
+    "    uint4 mask = get_sub_group_gt_mask();"
+    "    out[gid] = mask;\n"
+    "}\n";
+
+static const char *get_subgroup_le_mask_source =
+    "__kernel void test_get_sub_group_le_mask(const __global Type *in, "
+    "__global int4 *xy, __global Type *out)\n"
+    "{\n"
+    "    int gid = get_global_id(0);\n"
+    "    XY(xy,gid);\n"
+    "    xy[gid].z = get_max_sub_group_size();\n"
+    "    Type x = in[gid];\n"
+    "    uint4 mask = get_sub_group_le_mask();"
+    "    out[gid] = mask;\n"
+    "}\n";
+
+static const char *get_subgroup_lt_mask_source =
+    "__kernel void test_get_sub_group_lt_mask(const __global Type *in, "
+    "__global int4 *xy, __global Type *out)\n"
+    "{\n"
+    "    int gid = get_global_id(0);\n"
+    "    XY(xy,gid);\n"
+    "    xy[gid].z = get_max_sub_group_size();\n"
+    "    Type x = in[gid];\n"
+    "    uint4 mask = get_sub_group_lt_mask();"
+    "    out[gid] = mask;\n"
+    "}\n";
+
+static const char *get_subgroup_eq_mask_source =
+    "__kernel void test_get_sub_group_eq_mask(const __global Type *in, "
+    "__global int4 *xy, __global Type *out)\n"
+    "{\n"
+    "    int gid = get_global_id(0);\n"
+    "    XY(xy,gid);\n"
+    "    xy[gid].z = get_max_sub_group_size();\n"
+    "    Type x = in[gid];\n"
+    "    uint4 mask = get_sub_group_eq_mask();"
+    "    out[gid] = mask;\n"
+    "}\n";
+
+static const char *ballot_source =
+    "__kernel void test_sub_group_ballot(const __global Type *in, "
+    "__global int4 *xy, __global Type *out)\n"
+    "{\n"
+    "uint4 full_ballot = sub_group_ballot(1);\n"
+    "uint divergence_mask;\n"
+    "uint4 partial_ballot;\n"
+    "uint gid = get_global_id(0);"
+    "XY(xy,gid);\n"
+    "if (get_sub_group_local_id() & 1) {\n"
+    "    divergence_mask = 0xaaaaaaaa;\n"
+    "    partial_ballot = sub_group_ballot(1);\n"
+    "} else {\n"
+    "    divergence_mask = 0x55555555;\n"
+    "    partial_ballot = sub_group_ballot(1);\n"
+    "}\n"
+    " size_t lws = get_local_size(0);\n"
+    "uint4 masked_ballot = full_ballot;\n"
+    "masked_ballot.x &= divergence_mask;\n"
+    "masked_ballot.y &= divergence_mask;\n"
+    "masked_ballot.z &= divergence_mask;\n"
+    "masked_ballot.w &= divergence_mask;\n"
+    "out[gid] = all(masked_ballot == partial_ballot);\n"
+
+    "} \n";
+
+static const char *ballot_source_inverse =
+    "__kernel void test_sub_group_ballot_inverse(const __global "
+    "Type *in, "
+    "__global int4 *xy, __global Type *out)\n"
+    "{\n"
+    "    int gid = get_global_id(0);\n"
+    "    XY(xy,gid);\n"
+    "    Type x = in[gid];\n"
+    "    uint4 value = (uint4)(10,0,0,0);\n"
+    "    if (get_sub_group_local_id() & 1) {"
```

Comment:
The value returned by sub_group_inverse_ballot() does not depend on the uniformity of control flow within the subgroup. Why is divergence being tested here? It may be better not to test it without a clear idea as to the difference that it might contribute.

If it is necessary to test it, this should be done as per sub_group_elect() once the testing for that appropriately covers a maximum subgroup size of 128.

Reply:
It will by solved by https://github.com/KhronosGroup/OpenCL-CTS/issues/1205

Reply:
I think #1207 is the relevant issue for this comment.